### PR TITLE
feat: additionalClassSchema

### PR DESCRIPTION
**Pull request by dheimoz**
Sometimes you already have your css classes coupled and need to enforce them. Or in responsive design you have media queries coupled to some specific classes and adding a new class "schema-row" involves unnecessary complexity.

Therefore a with proposed prop additionalClassSchema in case it is passed the class can replace or complement the Div with "schema-row". In case it is not provided, it will render the default behaviour.

Hope this can help and expand the edge cases that your fantastic component is providing.

Should you need me to add tests, please let me know.

### CODE REVIEW

**Comment by marina-mosti** [comment]
@dheimoz thank you for taking the time to propose this change.

I see a couple of problems that we need to address. First of all, please make sure you run `yarn lint`, your personal linting and formatting settings are changing the whole file.

In terms of your idea, I find it usefull - however:

1. Let's rename the prop to `schemaRowClasses`.
1. Type the prop to `[String, Object, Array]` since any of those are valid ways to set CSS in Vue, I think we can then default to `null`.
1. Instead of toggling between `schema-row` and this property, let's make line 7 use both of them. `['schema-row', schemaRowClasses]`
1. Definitely need to have a unit case to ensure correct behavior :)
1. We need to document this new property too!

**Comment by dheimoz** [comment]
Great that you like it @marina-mosti will address the changes and document. Just wanted to make a proposal.

Keep you posted. I have to finish a project in 10 days (fortunately with Formvuelate will take less), after that I will provide the final revision.

**Comment by marina-mosti** [comment]
Thank you @dheimoz ping me when done :D

**Comment by dheimoz** [comment]
Hi @marina-mosti , as I have already made the changes I accommodate a couple of hours and wrote the documentation for the prop and a couple of tests.

As you might have guessed, English is not my natural language, so if anything sounds a little bit weird on the documentation, let me know and I can correct it.

Also, I wonder if I need to create more tests.

Thanks, I am in love with Formvuelate.

**Comment by marina-mosti** [comment]
@dheimoz sorry for the delay, Christmas and what nots cut through my time :) can you please make sure to Tun yarn lint on the files on your PR?

Theres a lot of lines being reformatted to whatever your IDE is set to. 

I'll review the docs as soon as I come back from vacation

**Comment by dheimoz** [comment]
Understood @marina-mosti have a great New Year and happy Vacations. They're deserved.

Will reformat and make some changes.

Thanks for the follow up.

File: index.md [hunk 1]
```diff
@@ -230,6 +230,75 @@ The example below applies a `margin-right` style to the first input.
   <<< .vitepress/docs/components/HorizontalForm.vue
 </SplitTab>
 
+### Prop: schemaRowClasses
+
+The philosophy is `Bring Your Own Component` and there will be moments when you already have your own css or a Framework UI you are using dictates that you need to use specific classes.
+
+For those opinated situations `schemaRowClasses` allows you to complement the `schema-row` class that Formvuelate provides.
+
+The prop accepts `[String, Object, Array]`, since any of those are valid ways to set CSS in Vue.
```
Comment (marina-mosti) [inline index.md]:
```suggestion
If you ever require to add additional classes to each structural row in the generated form, you can use the `schemaRowClasses` property to inject them.

These classes will be appended, and will _not_ substitute the `schema-row` class that FormVueLate already provides.

The prop accepts `[String, Object, Array]` as valid types.
```

File: index.md [hunk 2]
```diff
@@ -230,6 +230,75 @@ The example below applies a `margin-right` style to the first input.
...
+
+```html
+<SchemaForm
+  :schemaRowClasses="my-custom-class-a my-custom-class-b"
+  :schema="mySchema"
+/>
+```
+When the `Form` html element is rendered, there will be a `div` element with those classes:
```
Comment (marina-mosti) [inline index.md]:
```suggestion
Example output:
```

File: index.md [hunk 3]
```diff
@@ -230,6 +230,75 @@ The example below applies a `margin-right` style to the first input.
...
+  <div class="schema-row custom-class-a custom-class-b">
+    Fields
+  </div>
+</form>
+```
+:::tip
+You might find scenarios where the Field must have a class that depends on the class applied to `schemaRowClasses`. Fortunately, Formvuelate has you covered.
+
+For instance if you need this:
+```html
```
Comment (marina-mosti) [inline index.md]: This whole bit with the tip until the end is nice, but I feel it overexplains component composition which is not directly related to what FVL does. Let's remove this bit and keep it short and clean

**Review by marina-mosti — COMMENTED** [review]
Hey @dheimoz finally got around to reviewing, sorry for the delay. It's looking pretty good, I did some minor rewording of your docs to fit more with the writing style and left a few nitpicks that need to be addressed.

Let me know when you address those and get the coding style fixed and ill go ahead with the final review.

Thanks!

**Comment by dheimoz** [comment]
Hello @marina-mosti sorry for the late response. I am working on two projects. Will review the comments and correct as requested on weekend. Welcome back by the way.

**Comment by marina-mosti** [comment]
Hey @dheimoz no pressure, but are you still interested in pursuing this?

**Comment by dheimoz** [comment]
Yes, of course. Will work on this tomorrow. Thanks for the follow up.

**Comment by dheimoz** [comment]
Hello @marina-mosti , thank you very much for the patience. I have correct the linting problems and now only my contribution is listed. I have merged your valuable feedback on the documentation.

Looking forward your review!!

File: index.md [hunk 4]
```diff
@@ -230,6 +230,75 @@ The example below applies a `margin-right` style to the first input.
   <<< .vitepress/docs/components/HorizontalForm.vue
 </SplitTab>
 
+### Prop: schemaRowClasses
+
+If you ever require to add additional classes to each structural row in the generated form, you can use the `schemaRowClasses` property to inject them.
```
Comment (marina-mosti) [inline index.md]:
```suggestion
If you ever require to add additional classes to each structural row in the generated form, you can use the `schemaRowClasses` property of the `SchemaForm` component to inject them.
```

File: index.md [hunk 5]
```diff
@@ -230,6 +230,75 @@ The example below applies a `margin-right` style to the first input.
...
+
+These classes will be appended, and will _not_ substitute the `schema-row` class that FormVueLate already provides.
+
+The prop accepts `[String, Object, Array]` as valid types.
```
Comment (marina-mosti) [inline index.md]:
```suggestion
The `schemaRowClasses` prop accepts `[String, Object, Array]` as valid types, conforming to Vue's [class and style binding syntax](https://v3.vuejs.org/guide/class-and-style.html#binding-html-classes) 
```

File: index.md [hunk 6]
```diff
@@ -230,6 +230,75 @@ The example below applies a `margin-right` style to the first input.
...
+
+```html
+<SchemaForm
+  :schemaRowClasses="my-custom-class-a my-custom-class-b"
```
Comment (marina-mosti) [inline index.md]:
```suggestion
  schemaRowClasses="my-custom-class-a my-custom-class-b"
```

File: index.md [hunk 7]
```diff
@@ -230,6 +230,75 @@ The example below applies a `margin-right` style to the first input.
...
+  :schema="mySchema"
+/>
+```
+Example output:
```
Comment (marina-mosti) [inline index.md]: Please add a space between the closing 3 quotes in 246 and the next line

File: index.md [hunk 8]
```diff
@@ -230,6 +230,75 @@ The example below applies a `margin-right` style to the first input.
...
+
+```html
+<form>
+  <div class="schema-row custom-class-a custom-class-b">
+    Fields
```
Comment (marina-mosti) [inline index.md]:
```suggestion
    [...]
```

File: index.md [hunk 9]
```diff
@@ -230,6 +230,75 @@ The example below applies a `margin-right` style to the first input.
...
+        formData
+      }
+    }
+  }
+</script>
+```
+:::
+
+Your component (in this example `FormText`) must have the prop class and use it.
+
```
Comment (marina-mosti) [inline index.md]:
```suggestion
```

File: tests/unit/SchemaForm.spec.js [hunk 10]
```diff
@@ -17,6 +17,40 @@ markRaw(FormSelect)
...
+        component: FormText,
+        label: 'Last Name'
+      }
+    }
+    const wrapper = mount(SchemaForm, {
+      props: { schema, modelValue: {}, schemaRowClasses: '' }
+    })
+    expect(wrapper.find('.schema-row').exists()).toBe(true)
+  })
+})
```
Comment (marina-mosti) [inline tests/unit/SchemaForm.spec.js]: Please remove all this and insert the following after the closing `})` for the `it renders a form based on a schema` block

```
describe('schemaRowClasses prop', () => {
    it('renders form rows with user defined classes', () => {
      const schema = {
        firstName: {
          component: FormText,
          label: 'First Name'
        },
        lastName: {
          component: FormText,
          label: 'Last Name'
        }
      }
      
      const wrapper = mount(SchemaForm, {
        props: { schema, modelValue: {}, schemaRowClasses: 'custom-class-a' }
      })
      
      expect(wrapper.findAll('.schema-row + .custom-class-a').length()).toEqual(2)
    })
  })
```

**Review by marina-mosti — COMMENTED** [review]
Hey @dheimoz thanks for the changes, just a few more things to polish

File: tests/unit/SchemaForm.spec.js (hunk 10 above)
Comment (dheimoz) [reply tests/unit/SchemaForm.spec.js]: Understood. @marina-mosti working on them right now.

Comment (dheimoz) [reply tests/unit/SchemaForm.spec.js]: Hello @marina-mosti I have to inform that the test is failling
TypeError: wrapper.findAll(...).length is not a function

      54 |       })
      55 | 
    > 56 |       expect(wrapper.findAll('.schema-row + .custom-class-a').length()).toEqual(2)
         |                                                               ^
      57 |     })
      58 |   })
      59 | 
Will review the syntax and let you know.

Comment (marina-mosti) [reply tests/unit/SchemaForm.spec.js]: That's odd... https://vue-test-utils.vuejs.org/api/wrapper/#findall

Comment (marina-mosti) [reply tests/unit/SchemaForm.spec.js]: I think its a problem with the selector, should probably be `.schema-row.custom-class-a`

Comment (dheimoz) [reply tests/unit/SchemaForm.spec.js]: I have found this warning
https://vue-test-utils.vuejs.org/api/wrapper/#findall
Deprecation warning

Using findAll to search for Components is deprecated and will be removed. Use findAllComponents instead.

Comment (marina-mosti) [reply tests/unit/SchemaForm.spec.js]: You're not looking for a component, you're querying a class selector :)
`findAll` is basically doing `querySelectorAll`, which is correct. But the `+` in the CSS selector is incorrect. The classes are applied to the same element, not to a sybling

Comment (dheimoz) [reply tests/unit/SchemaForm.spec.js]: There is something wrong with VTU
      console.log(wrapper.html()) // Return HTML and clearly there are 2 div with both classes
      // https://vue-test-utils.vuejs.org/api/selectors.html
      console.log(wrapper.findAll('.schema-row').length); // Returns 2
      console.log(wrapper.findAll('.custom-class-a').length);// Returns 2
      console.log(wrapper.findAll('.schema-row + .custom-class-a').length); // Returns 1
      console.log(wrapper.findAll('.schema-row ~ .custom-class-a').length); // Returns 1

Comment (marina-mosti) [reply tests/unit/SchemaForm.spec.js]: As I said, it should be `.schema-row.custom-class-a`
Please update the test with the correct selector

File: index.md [hunk 11]
```diff
@@ -230,6 +230,30 @@ The example below applies a `margin-right` style to the first input.
...
+Example output:
+
+```html
+<form>
+  <div class="schema-row custom-class-a custom-class-b">
+    [...]
+  </div>
+</form>
+```
```
Comment (marina-mosti) [inline index.md:258]: Could you please also add a space after this bit?

File: tests/unit/SchemaForm.spec.js [hunk 12]
```diff
@@ -36,6 +36,27 @@ describe('SchemaForm', () => {
...
+          component: FormText,
+          label: 'Last Name'
+        }
+      }
+
+      const wrapper = mount(SchemaForm, {
+        props: { schema, modelValue: {}, schemaRowClasses: 'custom-class-a' }
+      })
+
+      expect(wrapper.findAll('.custom-class-a')).toHaveLength(2)
```
Comment (marina-mosti) [inline tests/unit/SchemaForm.spec.js]:
```suggestion
      expect(wrapper.findAll('.schema-row.custom-class-a')).toHaveLength(2)
```

File: index.md (hunk 11 above)
Comment (dheimoz) [reply index.md:258]: Done. I love your attention to this details. Have to improve personally for this.

**Comment by marina-mosti** [comment]
Thanks for the hard work @dheimoz :)
